docs/proposals/image-promotion.md

```diff
@@ -0,0 +1,240 @@
+images between different stages (eg. from *development* to *production*).
+This proposal also describes a way how to configure OpenShift to serve as
+a Continuous Delivery tool.
+
+## Constraints and Assumptions
+
+This document describes the following ways to promote images:
+
+* **Based on human intervention**
+  * using the `oc` and `docker` commands manually
+* **Automated by the OpenShift platform**
+  * using enhanced *DeploymentConfig* [lifecycle hooks](https://docs.openshift.org/latest/dev_guide/deployments.html#lifecycle-hooks)
+  * using [ImageChangeTrigger](https://docs.openshift.org/latest/dev_guide/deployments.html#image-change-trigger)
+  * using [Projects](https://docs.openshift.org/latest/dev_guide/projects.html) for different "stages"
+* **Based on external Continuous Delivery tool**
+  * [Jenkins](https://jenkins-ci.org)
+  * [go.cd](http://go.cd)
+  * [snap](http://snap-ci.com)
+
+Docker images can be promoted (and the use-cases are valid for):
+
+* **Within a single Project**
+* **Between multiple Projects**
+* **Between multiple Clusters**
+
+The Docker image promotion can be done using the Docker [image
+tags](https://docs.docker.com/userguide/dockerimages/#setting-tags-on-an-image)
+or Docker image [labels](https://docs.docker.com/userguide/labels-custom-metadata/).
+OpenShift also provides more features to add metadata, such as annotations or labels.
+Using them might allow users to develop complex conditional promotion scenarios, but
+this document does not describe them.
+
+In some cases, promoting just Docker images might not be enough and you want to promote
+the entire deployment or a project as a single unit of promotion. These scenarios
+are also not covered by this document.
+
+## Matrix
+
+* **Promotion based on human intervention**
+  * By creating ImageStreamTag manually
+    - `oc tag application/image:@sha256:02c104b application/image:qa-ready`
+  * By importing the Docker image manually using OpenShift CLI
+    - `oc import-image application/image`
+  * By moving the ImageStream from one project to another
+    - `oc export is -l name=application | oc create -n stage -f -`
+  * By using Docker CLI manually
+    - `docker tag application/image qa-registry:application/image:qa-ready`
+    - `docker push qa-registry:application/image:qa-ready`
+* **Automatic promotion**
+  * By using CLI
+    - Create cron job that tags the image when the image with watched tag is
+      available in registry
+  * By using OpenShift internal mechanism
+    - Use DeploymentConfig lifecycle hooks to do `oc tag`
+    - Use DeploymentConfig lifecycle hooks to do `docker tag` and `docker push`
+    - Use DeploymentConfig to call remote OpenShift API to do tagging
+    - Use DeploymentConfig strategy that deploys a 'testing' image and calls
+      the OpenShift API
+  * By using external Continuous delivery tools
+    - Use Jenkins job to do `oc tag` or call the OpenShift API
```

Comment:
"to do `oc tag`" sounds weird - maybe "Use a Jenkins job to invoke `oc tag` at the end of a successful build job - perhaps one that runs unit tests on a builder image created by another team"